pkg/bench/workload.go

```diff
@@ -194,6 +209,91 @@ func (w *writeWorkload) generateTimeSeries(id string, t time.Time) []prompb.Time
 	return timeseries
 }
 
+type batchReq struct {
+	batch   []prompb.TimeSeries
+	wg      *sync.WaitGroup
+	putBack chan []prompb.TimeSeries
+}
+
+func (w *writeWorkload) getSeriesBuffer(ctx context.Context) []prompb.TimeSeries {
+	select {
+	case <-ctx.Done():
+		return nil
+	case seriesBuffer := <-w.seriesBufferChan:
+		return seriesBuffer[:0]
+	}
+}
+
+func (w *writeWorkload) generateWriteBatch(ctx context.Context, id string, numBuffers int, seriesChan chan batchReq) error {
+	w.seriesBufferChan = make(chan []prompb.TimeSeries, numBuffers)
+	for i := 0; i < numBuffers; i++ {
+		w.seriesBufferChan <- make([]prompb.TimeSeries, 0, w.options.BatchSize)
+	}
+
+	seriesBuffer := w.getSeriesBuffer(ctx)
+	ticker := time.NewTicker(w.options.Interval)
+
+	defer close(seriesChan)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case timeNow := <-ticker.C:
+			now := timeNow.UnixNano() / int64(time.Millisecond)
+			wg := &sync.WaitGroup{}
+			for replicaNum := 0; replicaNum < w.replicas; replicaNum++ {
+				replicaLabel := prompb.Label{Name: "bench_replica", Value: fmt.Sprintf("replica-%05d", replicaNum)}
+				idLabel := prompb.Label{Name: "bench_id", Value: id}
+				for _, series := range w.series {
+					var value float64
+					switch series.seriesType {
+					case GaugeZero:
+						value = 0
+					case GaugeRandom:
+						value = rand.Float64()
+					case CounterOne:
+						value = series.lastValue + 1
+					case CounterRandom:
+						value = series.lastValue + float64(rand.Int())
+					default:
+						return fmt.Errorf("unknown series type %v", series.seriesType)
+					}
```

Comment:
this could fit nicely in it's own function

Reply:
I was considering it since this snippet is copied from the `generateTimeSeries` function above. However, I prefer keeping it in a single function.